### PR TITLE
refactor restore command

### PR DESCRIPTION
**Pull request by aawsome**

### CODE REVIEW

File: crates/rustic_core/src/commands/restore.rs [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+            let p = repo.pb.progress_spinner("setting metadata...");
+            self.restore_metadata(repo, dest, node)?;
+            p.finish();
+        }
+
+        Ok(())
+    }
+
+    /// collect restore information, scan existing files and remove superfluous files
+    pub(crate) fn collect_and_remove<P: ProgressBars, S: Indexed>(
```
[the remark this answers is not in the document]
Comment (aawsome) [reply crates/rustic_core/src/commands/restore.rs:97]: Yes, agree that the restore command is pretty complex on one hand and pretty optimized on the other hand...
I also thought about improving it, but then decided to do minimal changes in order to present the command in the API.

Can we work on the restore command in future?
Yes, it is very ugly that there is no possible repo method which would just analyze what needs to be done. But for instance regarding the removal of existing files, you have two alternatives:
- save the files to-remove in some kind of list (which can be very memory consuming if you delete a lot by this command)
- scan the destination (a the repository tree) a second time which can be quite costly.

I think refactoring this would need a bit more thinking and hence I don't want to do this parallel to the current task.

File: crates/rustic_core/src/commands/restore.rs [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+        dest.set_extended_attributes(path, &node.meta.extended_attributes)
+            .unwrap_or_else(|_| warn!("restore {:?}: setting extended attributes failed.", path));
+        dest.set_times(path, &node.meta)
+            .unwrap_or_else(|_| warn!("restore {:?}: setting file times failed.", path));
+    }
+}
+
+/// [`restore_contents`] restores all files contents as described by `file_infos`
+/// using the [`DecryptReadBackend`] `be` and writing them into the [`LocalBackend`] `dest`.
+fn restore_contents<P: ProgressBars, S: Open>(
```
[the remark this answers is not in the document]
Comment (aawsome) [reply crates/rustic_core/src/commands/restore.rs:350]: Wouldn't do this with this change. This function is basically just moved from the CLI crate where it was as this. At least now this function is very local within rustic_core and can be easily refactored.

**Comment by aawsome** [comment]
I added an example in the example dir.

File: crates/rustic_core/examples/restore.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    let repo = Repository::new(&repo_opts)
+        .unwrap()
+        .open()
+        .unwrap()
+        .to_indexed()
+        .unwrap();
+
+    // use latest snapshot without filtering snapshots
+    let node = repo.node_from_snapshot_path("latest", |_| true).unwrap();
```
[the remark this answers is not in the document]
Comment (aawsome) [reply crates/rustic_core/examples/restore.rs:26]: IMO, it is actually vice versa: When you want the `latest` snapshot, you usually specify extra conditions of w.r.t. what you want the latest snapshot.
On the other hand, if you specify a snapshots by (part of the) ID, you usually don't need any filtering as the snapshot is already defined.

I think the use cases how to get to a certain node which you want to restore will vary a lot depending on the user interface you are using. If you have a CLI, the `node_from_snapshot_path` is quite handy as it allows to give a string and will automatically resolve ids or latest snapshots from the string.
If you have some kind of GUI, I would prefer just fetching all snapshots using `repo.get_matching_snapshots()` or even `repo.get_all_snapshots()` (not present yet but would be `repo.get_matching_snapshots()` without filtering). Then a user could manually dive into some tree and pick the `Node` to be used for the restore.

TL;DR: Yes, we need to think about the API, but here I wouldn't do so ;-)
